File: 1364-tuple-with-same-product/tuple-with-same-product.go
func tupleSameProduct(nums []int) int {
    pairs := map[int]int{}
    ans, n := 0, len(nums)
    for i := 0; i < n; i++ {
        for j := i + 1; j < n; j++ {
            pairs[nums[i] * nums[j]]++
        }
    }
    for _, v := range pairs {
        ans += (v - 1) * v * 4
    }
    return ans
}

File: 1364-tuple-with-same-product/tuple-with-same-product.py
class Solution(object):
    def tupleSameProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        map = defaultdict(int)
        
        res = 0
        n = len(nums)
        for i in range(n):
            for j in range(i + 1, n):
                product = nums[i] * nums[j] 
                if product in map:
                    res += map[product] * 8
                map[product] += 1

        return res
